### PR TITLE
Experiment constructor

### PR DESCRIPTION
**Pull request by choisy**
I rewrote the experiment constructor. Following Hadley Wickham's recommandations regarding [S3 classes](https://adv-r.hadley.nz/s3.html#s3-classes) in the second edition of _Advanced R_, I have created

- the **constructor** `new_experiment()` that assembles an object of class `experiment` from all its components provided individually. This function performs a number of tests on the structure of the elements. This is an internal function;
- the **validator** `validate_experiment()` that takes an object of class `experiment` as an input and performs a number of tests on it, regarding the content of the data, rather than the structure of the data (which is tested in the constructor). Note in particular, that I have also written 3 new functions `get_parameters_names()`, `get_variables_names()` and `get_all_names()` (all in the `get_names.R` file) that allow to retrieve parameters and variables names of a model by directly parsing the GAML file (i.e. without having to start java and GAMA as `read_gaml_experiment()` does (see below about this new function) and that takes forever). The validator returns its input and is also an internal function;
- the **helper** `experiment()` which is exported and has the same interface as the constructor and which basically calls successively the constructor and the validator. It thus returns an object of class `experiment`.

With this done, I updated

- `load_experiment()` that now calls the helper `experiment()`. Note also that in the process I have isolated the piece of code from this function that reads the GAML file and used it to make the function `read_gaml_experiment()` that I included in `utils.R`. Still on `load_experiment()`, I have also de-duplicated the piece of code that extracted the information on parameters values and periods of observation. This eases the code maintenance and makes it less prone to bugs;
- `map_experiment()` that maps a data frame onto an object of class `experiment` and that now calls the validator `validate_experiment()`;
- `as_experiment()` that converts a data frame into an  `experiment` and that now calls the helper `experiment()`.

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -58,15 +55,13 @@ importFrom(plot3D,scatter3D)
...
 importFrom(purrr,pmap)
-importFrom(stats,na.omit)
 importFrom(stats,setNames)
 importFrom(stats,var)
 importFrom(stringr,regex)
 importFrom(stringr,str_detect)
 importFrom(stringr,str_match)
 importFrom(stringr,str_match_all)
 importFrom(tibble,as_tibble)
-importFrom(utils,capture.output)
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply NAMESPACE]: Done in commit `ea71886`

File: R/experiment.R [hunk 2]
```diff
@@ -1,283 +1,114 @@
...
+  names_obsrates <- names(obsrates)
+  parvarnames <- c(paste0("p_", names_param), paste0("r_", names_obsrates))
+
+  if (is.null(dic))
+    dic <- setNames(c(parvarnames, "tmax", "seed"),
+                    c(names_param, names_obsrates, "tmax", "seed"))
+
+  obsrates[] <- lapply(obsrates, as.integer)
+  structure(setNames(cbind(parameters, obsrates, tmax = as.integer(tmax), seed = seed), dic),
+            class      = c("experiment", "tbl_df", "tbl", "data.frame"),
```
[the remark this answers is not in the document]
Comment (choisy) [reply R/experiment.R]: Thanks.

File: NAMESPACE (hunk 1 above)
[the remark this answers is not in the document]
Comment (choisy) [reply NAMESPACE]: Thanks.

File: R/experiment.R [hunk 3]
```diff
@@ -1,283 +1,114 @@
...
-}
+  if (length(setdiff(colnames, dic)) > 0)
+    stop("Some variables or parameters names are not in the dictionary.")
 
+  if (setequal(dic, names(dic_rev)) + setequal(names(dic), dic_rev) < 2)
+    stop("The dictionaries are inconsistent.")
 
+  if (length(setdiff(sub("^[pr]_", "", colnames), get_all_names(model))) > 0)
+    stop(paste("The names of parameters and observed variables in experiment",
+               substitute(x), "do not correspond to the names of parameters and",
```
[the remark this answers is not in the document]
Comment (LucieContamin) [reply R/experiment.R]: For example, if with the orginial version I have:
```
> df <-   data.frame(S0 = c(900, 800, 500),
+                   I0 = c(100, 200, 500),
+                   R0 = 0,
+                   beta = 1.5,
+                   gama = .15,
+                   S = c(1, 2, 3),
+                   I = c(2, 4, 6),
+                   R = c(10, 20, 30),
+                   nbiter = 1000,
+                   seed = "123456789")
> exp0 <- as_experiment(df, parameters = c("S0", "I0", "R0", "beta", "gama"),
+                       obsrates  = c("S", "I", "R"),
+                       tmax = "nbiter", seed = "seed", experiment = "sir",
+                       model =
+                        system.file("models", "sir.gaml", package = "rama"))
# Using default directory name " sir " in current directory " /Users/mac/Desktop/rama ".
# Simulations results will be saved in " sir ".
# Error in validate_experiment(new_experiment(parameters, obsrates, tmax,  : 
#   The names of parameters and observed variables in experiment new_experiment do not correspond to the names of parameters and variables defined in the /Library/Frameworks/R.framework/Versions/3.5/Resources/library/rama/models/sir.gaml file.
# The names of parameters and observed variables in experiment parameters do not correspond to the names of parameters and variables defined in the /Library/Frameworks/R.framework/Versions/3.5/Resources/library/rama/models/sir.gaml file. 
# The names of parameters and observed variables in experiment obsrates do not correspond to the names of parameters and variables defined in the /Library/Frameworks/R.framework/Versions/3.5/Resources/library/rama/models/sir.gaml file.
# The names of parameters and observed variables in experiment tmax do not correspond to the names of parameters and variables defined in the /Library/Frameworks/R.framework/Versions/3.5/Resources/library/rama/models/sir.gaml file.
# The names of parameters and observed variables in expe...
```

and with the second version I have:

```
> df <-   data.frame(S0 = c(900, 800, 500),
+                   I0 = c(100, 200, 500),
+                   R0 = 0,
+                   beta = 1.5,
+                   gama = .15,
+                   S = c(1, 2, 3),
+                   I = c(2, 4, 6),
+                   R = c(10, 20, 30),
+                   nbiter = 1000,
+                   seed = "123456789")
> exp0 <- as_experiment(df, parameters = c("S0", "I0", "R0", "beta", "gama"),
+                       obsrates  = c("S", "I", "R"),
+                       tmax = "nbiter", seed = "seed", experiment = "sir",
+                       model =
+                        system.file("models", "sir.gaml", package = "rama"))
# Using default directory name " sir " in current directory " /Users/mac/Desktop/rama ".
# Simulations results will be saved in " sir_1 ".
# Error in validate_experiment(new_experiment(parameters, obsrates, tmax,  : 
# The names of parameters and observed variables in experiment gama do not correspond to the names of parameters and variables defined in the sir.gaml file.
```

File: R/experiment.R [hunk 4]
```diff
@@ -1,283 +1,123 @@
...
-  obs <- na.omit(stringr::str_match(names(exp), "r_(.*)")[, 2])
-  experiment(parameters = params,
-             obsrates = obs,
-             tmax = "tmax",
-             seed = "seed",
-             experiment = name(exp),
-             model = model(exp),
-             dir = dir,
-             df = df)
+#'
```
[the remark this answers is not in the document]
Comment (choisy) [reply R/experiment.R]: Yep! Done in commit a422bf151284cfcf1ead734e470bd15f232559b5

File: R/experiment.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (choisy) [reply R/experiment.R]: Got it! Thanks a lot for the example.
Done in commit 91424a30da068d79451780797743a8417d90e6ca

**Pull request merged**